### PR TITLE
[ML] Reset memory status from soft limit to ok below threshold

### LINKED ISSUE #1131: [ML] Discrepancy in resetting of soft_limit status
If an anomaly detection job is approaching its memory limit then we start pruning old models more aggressively and the memory status is set to `soft_limit`.  If the memory usage then reduces such that aggressive pruning is no longer required then we do not immediately reset the memory status to `ok`.  However, if the job is closed and reopened, or if it switches nodes because the node it was running on leaves the cluster then the memory status _is_ reset to `ok` if pruning is no longer required.

This discrepancy should be addressed.  If aggressive pruning is no longer required then the job's memory status should be reset to `ok` without requiring a process restart.

### PR DESCRIPTION
**Pull request by droberts195**
We start pruning models when the memory used goes above the
prune threshold, and set the job's memory status to soft
limit at the same time.  If you close and reopen the job and
memory usage is lower than the prune threshold then we don't
prune and the memory limit is ok.  Prior to this change there
was a discrepancy where if you didn't close and reopen the
job but memory usage dropped below the prune threshold then
we continued to prune and left the memory status in soft
limit.

This change makes the case of not closing and reopening the
job more consistent.  If memory usage drops below the prune
threshold and the prune window expands to the default level
then we stop pruning and set the memory status back to ok.

Closes #1131

### CODE REVIEW

File: include/model/CResourceMonitor.h [hunk 1]
```diff
@@ -127,12 +127,18 @@ class MODEL_EXPORT CResourceMonitor {
     //! We are being told that aggressive pruning has taken place
     //! to avoid hitting the resource limit, and we should report this
     //! to the user when we can
-    void acceptPruningResult();
+    void acceptPruningStartResult();
```
Comment (tveasey) [inline include/model/CResourceMonitor.h:130]: This matched what we had before, but I find these names rather strange. How about:
```suggestion
    void startPruning();
```

File: lib/model/CResourceMonitor.cc [hunk 2]
```diff
@@ -187,11 +187,21 @@ bool CResourceMonitor::pruneIfRequired(core_t::TTime endTime) {
...
+                    // If we increase the prune window to a size that's bigger
+                    // than what we started with then we should stop pruning to
+                    // be consistent with what would happen if the job was
+                    // closed and reopened
+                    LOG_DEBUG(<< "Pruning no longer necessary.");
+                    m_PruneWindow = m_PruneWindowMaximum;
+                    m_HasPruningStarted = false;
+                    this->acceptPruningEndResult();
+                } else {
+                    LOG_TRACE(<< "Expanding window, to " << m_PruneWindow);
```
Comment (tveasey) [inline lib/model/CResourceMonitor.cc:203]: nit:
```suggestion
                    LOG_TRACE(<< "Expanding window to " << m_PruneWindow);
```

File: include/model/CResourceMonitor.h (hunk 1 above)
Comment (droberts195) [reply include/model/CResourceMonitor.h:130]: `startPruning()` sort of implies that it's going to do more than maybe update the memory status.  So I renamed it but moved the debug and setting of `m_HasPruningStarted` into this method too.  And the equivalent for `endPruning()`.  Does that sound OK to you?  I am just testing it locally now.

Comment (tveasey) [reply include/model/CResourceMonitor.h:130]: That sounds much better to me

**Comment by droberts195** [comment]
retest

**Pull request merged**